File: src/monitor.py
"""
Clipboard monitoring and capture logic.

Tracks system clipboard changes, captures text and images, and stores 
them in the database. Includes logic to avoid re-capturing content 
that the daemon itself just pasted.
"""
import hashlib
import time
import uuid
import os
from pathlib import Path

import gi
gi.require_version('Gtk', '4.0')
gi.require_version('Gdk', '4.0')
from gi.repository import Gdk, GLib

IMAGES_DIR = Path.home() / '.local' / 'share' / 'clipboard-history' / 'images'


class ClipboardMonitor:
    """Watchdog for clipboard changes."""

    def __init__(self, db):
        """Initialize with a reference to the history database."""
        self.db = db
        self._skip_hash: str | None = None  # hash of content set by the daemon
        self._reading = False

    def start(self):
        """Register for clipboard change signals."""
        display = Gdk.Display.get_default()
        if display is None:
            return
        self._clipboard = display.get_clipboard()
        self._clipboard.connect('changed', self._on_changed)

    def set_skip_hash(self, hash_: str):
        """Pre-register a hash to ignore on the next change event (pasting)."""
        self._skip_hash = hash_

    def _on_changed(self, clipboard):
        """Signal handler for clipboard 'changed' event."""
        if clipboard.is_local():
            return
        if self._reading:
            return
        self._reading = True
        # Prioritize text capture
        clipboard.read_text_async(None, self._on_text_ready)

    def _on_text_ready(self, clipboard, result):
        """Callback for text data retrieval."""
        try:
            text = clipboard.read_text_finish(result)
        except Exception:
            text = None

        if text and text.strip():
            h = hashlib.sha256(text.encode()).hexdigest()
            # Verify if this was a self-inflicted change (paste)
            if h != self._skip_hash:
                self.db.add_entry('text', content=text)
            self._skip_hash = None
            self._reading = False
        else:
            # No text or skipped - try image capture
            self._skip_hash = None
            formats = clipboard.get_formats()
            mime_types = list(formats.get_mime_types())
            if any('image' in m for m in mime_types):
                clipboard.read_texture_async(None, self._on_texture_ready)
            else:
                self._reading = False

    def _on_texture_ready(self, clipboard, result):
        """Callback for image texture retrieval."""
        try:
            texture = clipboard.read_texture_finish(result)
        except Exception:
            texture = None
        finally:
            self._reading = False

        if texture is None:
            return

        try:
            # Convert texture to PNG bytes
            png_bytes = texture.save_to_png_bytes()
            raw = bytes(png_bytes.get_data())
            h = hashlib.sha256(raw).hexdigest()
            img_path = IMAGES_DIR / f'{h[:16]}.png'
            
            if not img_path.exists():
                # Save full image first
                texture.save_to_png(str(img_path))
                # Optional: Generate thumbnail to save space/memory in UI
                try:
                    from PIL import Image as PILImage
                    with PILImage.open(str(img_path)) as im:
                        im.thumbnail((480, 320))
                        im.save(str(img_path))
                except Exception:
                    pass  # Fallback to full-res if Pillow is missing
            self.db.add_entry('image', image_path=str(img_path), raw=raw)
        except Exception as e:
            print(f'[monitor] image error: {e}')

File: src/monitor.rs
use crate::db::{hash_bytes, Database};
use gtk::gdk;
use gtk::prelude::*;
use std::cell::{Cell, RefCell};
use std::fs;
use std::io::Write;
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::rc::Rc;

pub struct ClipboardMonitor {
    clipboard: gdk::Clipboard,
    db: Rc<Database>,
    reading: Cell<bool>,
    skip_hash: RefCell<Option<String>>,
}

impl ClipboardMonitor {
    pub fn start(db: Rc<Database>) -> Option<Rc<Self>> {
        let display = gdk::Display::default()?;
        let clipboard = display.clipboard();
        let monitor = Rc::new(Self {
            clipboard,
            db,
            reading: Cell::new(false),
            skip_hash: RefCell::new(None),
        });

        let weak = Rc::downgrade(&monitor);
        monitor.clipboard.connect_changed(move |_| {
            if let Some(monitor) = weak.upgrade() {
                monitor.on_changed();
            }
        });

        Some(monitor)
    }

    pub fn set_text(&self, text: &str) {
        *self.skip_hash.borrow_mut() = Some(hash_bytes(text.as_bytes()));
        self.clipboard.set_text(text);
        wl_copy("text/plain", text.as_bytes());
    }

    pub fn set_image_from_path(&self, path: &str) -> bool {
        match gdk::Texture::from_filename(path) {
            Ok(texture) => {
                if let Ok(bytes) = fs::read(path) {
                    *self.skip_hash.borrow_mut() = Some(hash_bytes(&bytes));
                    wl_copy("image/png", &bytes);
                }
                self.clipboard.set_texture(&texture);
                true
            }
            Err(error) => {
                eprintln!("[monitor] failed to load image for paste: {error:#}");
                false
            }
        }
    }

    fn on_changed(self: &Rc<Self>) {
        if self.clipboard.is_local() || self.reading.replace(true) {
            return;
        }

        let weak = Rc::downgrade(self);
        self.clipboard
            .read_text_async(None::<&gio::Cancellable>, move |result| {
                if let Some(monitor) = weak.upgrade() {
                    monitor.on_text_ready(result);
                }
            });
    }

    fn on_text_ready(self: &Rc<Self>, result: Result<Option<glib::GString>, glib::Error>) {
        let Ok(Some(text)) = result else {
            *self.skip_hash.borrow_mut() = None;
            self.read_texture();
            return;
        };

        let text = text.to_string();
        if text.trim().is_empty() {
            *self.skip_hash.borrow_mut() = None;
            self.read_texture();
            return;
        }

        let hash = hash_bytes(text.as_bytes());
        if self.skip_hash.borrow().as_deref() != Some(hash.as_str()) {
            if let Err(error) = self.db.add_text(&text) {
                eprintln!("[monitor] failed to store clipboard text: {error:#}");
            }
        }
        *self.skip_hash.borrow_mut() = None;
        self.reading.set(false);
    }

    fn read_texture(self: &Rc<Self>) {
        let weak = Rc::downgrade(self);
        self.clipboard
            .read_texture_async(None::<&gio::Cancellable>, move |result| {
                if let Some(monitor) = weak.upgrade() {
                    monitor.on_texture_ready(result);
                }
            });
    }

    fn on_texture_ready(self: &Rc<Self>, result: Result<Option<gdk::Texture>, glib::Error>) {
        self.reading.set(false);

        let Ok(Some(texture)) = result else {
            return;
        };

        let png_bytes = texture.save_to_png_bytes();
        let bytes = png_bytes.as_ref();
        let hash = hash_bytes(bytes);
        if self.skip_hash.borrow().as_deref() == Some(hash.as_str()) {
            *self.skip_hash.borrow_mut() = None;
            return;
        }

        let path: PathBuf = self.db.images_dir().join(format!("{}.png", &hash[..16]));
        if !path.exists() {
            if let Err(error) = texture.save_to_png(&path) {
                eprintln!("[monitor] failed to save clipboard image: {error:#}");
                return;
            }
        }

        if let Some(path) = path.to_str() {
            if let Err(error) = self.db.add_image(path, bytes) {
                eprintln!("[monitor] failed to store clipboard image: {error:#}");
            }
        }
        *self.skip_hash.borrow_mut() = None;
    }
}

fn wl_copy(mime_type: &str, bytes: &[u8]) {
    let Ok(mut child) = Command::new("wl-copy")
        .args(["--type", mime_type])
        .stdin(Stdio::piped())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn()
    else {
        return;
    };

    if let Some(stdin) = child.stdin.as_mut() {
        let _ = stdin.write_all(bytes);
    }
}
